CovidWatch/app/src/main/java/org/covidwatch/android/ble/BluetoothManager.kt

```diff
@@ -0,0 +1,181 @@
+import org.covidwatch.android.data.ContactEvent
+import org.covidwatch.android.data.ContactEventDAO
+import org.covidwatch.android.data.CovidWatchDatabase
+import org.covidwatch.libcontactrace.BluetoothService
+import org.covidwatch.libcontactrace.BluetoothService.LocalBinder
+import org.covidwatch.libcontactrace.cen.*
+import org.covidwatch.libcontactrace.toBytes
+import org.covidwatch.libcontactrace.toUUID
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+interface BluetoothManager {
+    fun startAdvertiser(cen: Cen)
+    fun stopAdvertiser()
+
+    fun startService(cen: Cen)
+    fun stopService()
+
+    fun changeAdvertisedValue(cen: Cen)
+}
+
+class BluetoothManagerImpl(
+    private val app: Application
+) : BluetoothManager {
+
+    private val intent get() = Intent(app, BluetoothService::class.java)
+
+    private var service: BluetoothService? = null
+
+    private val cenGenerator = DefaultCenGenerator()
+    private val cenVisitor = DefaultCenVisitor(app)
+    private var timer: Timer? = null
+
+    private val serviceConnection: ServiceConnection = object : ServiceConnection {
+        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
+            this@BluetoothManagerImpl.service = (service as LocalBinder).service.apply {
+                configure(
+                    BluetoothService.ServiceConfiguration(
+                        cenGenerator,
+                        cenVisitor,
+                        foregroundNotification()
+                    )
+                )
+                start()
+            }
+
+            runTimer()
+        }
+
+        override fun onServiceDisconnected(name: ComponentName?) = Unit
+    }
+
+    private fun runTimer() {
+        // scheduler a new timer to start changing the contact event numbers
+        timer?.cancel()
+        timer = Timer()
+        timer?.scheduleAtFixedRate(
+            object : TimerTask() {
+                override fun run() {
+                    service?.updateCen()
+                }
+            },
+            TimeUnit.MINUTES.toMillis(CEN_CHANGE_INTERVAL_MIN),
+            TimeUnit.MINUTES.toMillis(CEN_CHANGE_INTERVAL_MIN)
+        )
+    }
+
+    private fun foregroundNotification(): Notification {
+        createNotificationChannelIfNeeded()
+
+        val notificationIntent = Intent(app, MainActivity::class.java)
+        val pendingIntent = PendingIntent.getActivity(
+            app, 0, notificationIntent, PendingIntent.FLAG_UPDATE_CURRENT
+        )
+
+        return NotificationCompat.Builder(app, CHANNEL_ID)
+            .setContentTitle("Tags is logging")
+            .setSmallIcon(R.drawable.ic_launcher_foreground)
+            .setContentIntent(pendingIntent)
+            .setCategory(Notification.CATEGORY_SERVICE)
+            .build()
+    }
+
+    override fun changeAdvertisedValue(cen: Cen) {
+        cenGenerator.cen = cen
+        service?.updateCen()
+    }
+
+    override fun startAdvertiser(cen: Cen) {
```

Comment:
so the `BluetoothService` that you start at line 59 also starts the advertiser and scanner, right? So why is a separate `startAdvertiser` needed?

Reply:
This is, part of the compatibility with CoEpi code that I tried to maintain, that had this method in their interface so I added this functionality. But they seem to not use it :)